commit 5cf882d8b74747bbc08463d83cf80509c920edca
Author: borislav nikolov <jack@sofialondonmoskva.com>
Date:   Sat Mar 21 23:42:22 2020 +0100

    add copy_selection_to_user

diff --git a/drivers/tty/vt/selection.c b/drivers/tty/vt/selection.c
index d54a549c5892..9b26dec762dd 100644
--- a/drivers/tty/vt/selection.c
+++ b/drivers/tty/vt/selection.c
@@ -6,6 +6,7 @@
  *			       struct tty_struct *)'
  *     'int set_selection_kernel(struct tiocl_selection *, struct tty_struct *)'
  *     'void clear_selection(void)'
+ *     'int copy_selection_to_user(char __user *)'
  *     'int paste_selection(struct tty_struct *)'
  *     'int sel_loadlut(char __user *)'
  *
@@ -71,6 +72,45 @@ sel_pos(int n, bool unicode)
 	return inverse_translate(vc_sel.cons, screen_glyph(vc_sel.cons, n), 0);
 }
 
+/**
+ *	copy_selection_to_user		-	get current selection
+ *
+ *	Get a copy of current selection, console lock does not have to
+ *	be held
+ */
+int copy_selection_to_user(char __user *arg)
+{
+	int get_sel_user_size;
+	int ret;
+
+	if (copy_from_user(&get_sel_user_size,
+			   arg,
+			   sizeof(vc_sel.buf_len)))
+		return -EFAULT;
+
+	mutex_lock(&vc_sel.lock);
+
+	if (get_sel_user_size < vc_sel.buf_len) {
+
+		mutex_unlock(&vc_sel.lock);
+
+		return -EFAULT;
+	}
+
+	ret = copy_to_user(arg,
+			   &vc_sel.buf_len,
+			   sizeof(vc_sel.buf_len));
+	if (ret == 0)
+		ret = copy_to_user(arg+sizeof(vc_sel.buf_len),
+				   vc_sel.buffer,
+				   vc_sel.buf_len);
+
+	mutex_unlock(&vc_sel.lock);
+
+	return ret;
+}
+EXPORT_SYMBOL_GPL(copy_selection_to_user);
+
 /**
  *	clear_selection		-	remove current selection
  *
diff --git a/drivers/tty/vt/vt.c b/drivers/tty/vt/vt.c
index 309a39197be0..2b7eb55aafa3 100644
--- a/drivers/tty/vt/vt.c
+++ b/drivers/tty/vt/vt.c
@@ -3061,6 +3061,9 @@ int tioclinux(struct tty_struct *tty, unsigned long arg)
 		case TIOCL_PASTESEL:
 			ret = paste_selection(tty);
 			break;
+		case TIOCL_GETSEL:
+			ret = copy_selection_to_user(p+1);
+			break;
 		case TIOCL_UNBLANKSCREEN:
 			console_lock();
 			unblank_screen();
diff --git a/include/linux/selection.h b/include/linux/selection.h
index 5b890ef5b59f..7cb971795013 100644
--- a/include/linux/selection.h
+++ b/include/linux/selection.h
@@ -15,6 +15,7 @@ struct tty_struct;
 struct vc_data;
 
 extern void clear_selection(void);
+extern int copy_selection_to_user(char __user *arg);
 extern int set_selection_user(const struct tiocl_selection __user *sel,
 			      struct tty_struct *tty);
 extern int set_selection_kernel(struct tiocl_selection *v,
diff --git a/include/uapi/linux/tiocl.h b/include/uapi/linux/tiocl.h
index b32acc229024..055ebda041d4 100644
--- a/include/uapi/linux/tiocl.h
+++ b/include/uapi/linux/tiocl.h
@@ -20,6 +20,7 @@ struct tiocl_selection {
 };
 
 #define TIOCL_PASTESEL	3	/* paste previous selection */
+#define TIOCL_GETSEL	18	/* get current selection */
 #define TIOCL_UNBLANKSCREEN	4	/* unblank screen */
 
 #define TIOCL_SELLOADLUT	5
